jetty-core/jetty-quic/jetty-quic-api/src/main/java/org/eclipse/jetty/quic/api/Stream.java

```diff
@@ -0,0 +1,428 @@
+     * <p>Returns whether the stream has been created locally or remotely.</p>
+     *
+     * @return whether the stream is local or remote
+     */
+    boolean isLocal();
+
+    /**
+     * <p>Returns whether the stream is fully closed,
+     * both locally and remotely.</p>
+     *
+     * @return whether the stream is fully closed
+     * @see #isLocallyClosed()
+     * @see #isRemotelyClosed()
+     */
+    boolean isClosed();
+
+    /**
+     * <p>Returns whether the stream is locally closed.</p>
+     * <p>A locally closed stream will not send further data.</p>
+     * <p>A stream becomes locally closed when either it has sent
+     * {@link #data(boolean, List, Promise.Invocable) the last data}, or issued
+     * a {@link #reset(long, Promise.Invocable)}.</p>
+     *
+     * @return whether the stream is locally closed
+     * @see #isRemotelyClosed()
+     * @see #isClosed()
+     */
+    boolean isLocallyClosed();
+
+    /**
+     * <p>Returns whether the stream is remotely closed.</p>
+     * <p>A remotely closed stream will not receive further data.</p>
+     * <p>A stream becomes remotely closed when it has {@link #read()}
+     * the last data.</p>
+     *
+     * @return whether the stream is remotely closed
+     * @see #isLocallyClosed()
+     * @see #isClosed()
+     */
+    boolean isRemotelyClosed();
+
+    long getIdleTimeout();
+
+    void setIdleTimeout(long idleTimeout);
+
+    /**
+     * <p>Returns the {@link Session} this stream is associated to.</p>
+     *
+     * @return the {@link Session} this stream is associated to
+     */
+    Session getSession();
+
+    /**
+     * <p>Reads data from this stream, if any.</p>
+     * <p>The returned {@link Data} object may be {@code null} indicating
+     * that the end of the read side of the stream has not yet been reached,
+     * which may happen when not all the bytes have been received so far,
+     * for example the remote peer did not send them yet, or they are
+     * in-flight.</p>
+     * <p>When the returned {@link Data} object is not {@code null},
+     * the flow control window has been enlarged by the data length.
+     * Applications <em>must</em> call, either immediately or later (even
+     * asynchronously from a different thread) {@link Data#release()}
+     * to notify the implementation that the bytes have been processed.</p>
+     * <p>{@link Data} objects may be stored away for later,
+     * asynchronous, processing (for example, to process them only when
+     * all of them have been received).</p>
+     * <p>Once the returned {@link Data} object indicates that the end
+     * of the read side of the stream has been reached, further calls to this
+     * method will return a {@link Data} object with the same indication,
+     * although the instance may be different.</p>
+     *
+     * @return a {@link Data} object containing the data bytes,
+     * or {@code null} if no data bytes are available
+     * @see #demand()
+     */
+    Data read();
```

Comment:
Why is there no way to return errors from this method? I think we benefited a lot from the `Content.Source#read` contract, which can serialize failures with data arrival.  What happens if a call to `read` is racing with a call to `Listener#onFailure`?

I also see there is a `StreamDataFailure` record, that does have a failure in it, but nothing in the contract of read says to expect that might be returned?

Reply:
Converted to return `Chunk`.